### PR TITLE
READY: Channels GUI changes

### PR DESCRIPTION
**Pull request by ichorid**
This PR contains extensive changes to Channels tables layout. It adds toggle buttons instead of "subscribe :heart:" , adds progress bars to updating channels, adds emojis to categories, etc.

After this PR, I'll file a couple of follow-ups to keep the other parts of the Channels GUI in line with the style introduced in this one. Everyone is very welcome to just use Tribler and file your comments about necessary style and interaction changes, as GUIs are complex mazes of interactive connections, and it is impossible to make everything proper in one go.

Known issues:
* Channel indicator on the channel screen (right corner) is visualised differently than the one in the table
* Progress bars on torrents in the personal channel are not updated dynamically
* On Mac, there is some kind of race condition in GUI tests which prevents the test from running correctly (hope someone with a Mac will help me on this :wink: )
* Some toolips and icons may be missing (please report on this)

discov

image

### CODE REVIEW

**Comment by ichorid** [comment]
retest this please

File: src/tribler-core/tribler_core/tests/test_notifier.py [hunk 1]
```diff
@@ -1,14 +1,20 @@
...
-    def callback_func(*_):
-        callback_func.called = True
-
-    notifier.add_observer(NTFY.TORRENT_FINISHED, callback_func)
+    mock_foo = Mock()
+    notifier.add_observer(NTFY.TORRENT_FINISHED, mock_foo.bar)
     notifier.notify(NTFY.TORRENT_FINISHED)
-    assert callback_func.called
+    # Notifier uses asyncio loop internally, so we must wait at least a single loop cycle
+    await deliver_messages()
```
Comment (devos50) [inline src/tribler-core/tribler_core/tests/test_notifier.py:19]: I'm not sure if `deliver_messages` is the best way to skip a event loop cycle (it also looks that you are importing it from the tunnel tests?). Are there any alternatives provided by asyncio to achieve this?

File: src/tribler-gui/tribler_gui/widgets/channelcontentswidget.py [hunk 2]
```diff
@@ -87,6 +83,7 @@ def __exit__(self, *args):
                     obj.blockSignals(False)
 
         self.freeze_controls = freeze_controls_class
+        self.setStyleSheet("QToolTip { color: #222222; background-color: #eeeeee; border: 0px; }")
```
Comment (devos50) [inline src/tribler-gui/tribler_gui/widgets/channelcontentswidget.py:86]: Would it work when we apply this style sheet on a top-level widget (e.g., the window)? I assume that we are not going to use different styles for the `QToolTip`.

**Review by devos50 — CHANGES_REQUESTED** [review]
It's a bit hard to review GUI code since I cannot easily predict what Qt will render :). So I also executed your PR. I noticed the following things (on Mac):

Schermafbeelding 2020-10-21 om 10 20 53

- There seems to be some negative margin on the left of the table, and around 50 pixels of unused space to the right of the table.
- The globe emoji has a different color scheme than the one in the left menu. You could replace it with the figure maybe?
- There is (too) much horizontal padding between the vertical lines in the popularity indicator.
- The tooltip when hovering over the popularity indicator indicates a percentage, which is not intuitive for me. What does 9% mean? Suggestion: use language like 'unpopular', 'popular', 'very populary'.

Schermafbeelding 2020-10-21 om 10 23 46

- The toggled button looks off.

Schermafbeelding 2020-10-21 om 10 25 21

- I think that re-using the popularity indicator here, in a different color, is not clear. Why not re-use the toggle?

Schermafbeelding 2020-10-21 om 10 27 35

- The progress bar when downloading does not render correctly.

File: src/tribler-core/tribler_core/tests/test_notifier.py (hunk 1 above)
Comment (qstokkink) [reply src/tribler-core/tribler_core/tests/test_notifier.py:19]: > Are there any alternatives provided by asyncio to achieve this?

If you only want to yield the current execution loop, you can use `asyncio.sleep(0)`.

File: src/tribler-core/tribler_core/modules/metadata_store/gigachannel_manager.py [hunk 3]
```diff
@@ -285,14 +285,15 @@ async def download_channel(self, channel):
...
+                channel_dirname = self.session.mds.get_channel_dir_path(channel)
                 self.session.mds.process_channel_dir(
                     channel_dirname, channel.public_key, channel.id_, external_thread=True
                 )
-                self.session.mds._db.disconnect()
             except Exception as e:
                 self._logger.error("Error when processing channel dir download: %s", e)
                 return
+            finally:
+                self.session.mds._db.disconnect()
```
[the remark this answers is not in the document]
Comment (ichorid) [reply src/tribler-core/tribler_core/modules/metadata_store/gigachannel_manager.py:296]: As the method name implies, we do it in an external thread. Pony does not automatically disconnect external thread connections on thread exit, so it is a necessity.

File: src/tribler-core/tribler_core/modules/metadata_store/restapi/channels_endpoint.py [hunk 4]
```diff
@@ -79,7 +83,31 @@ async def get_channels(self, request):
...
-            channels_list = [channel.to_simple_dict() for channel in channels]
+            channels_list = []
+            for channel in channels:
+                channel_dict = channel.to_simple_dict()
+                # Add progress info for those channels that are still being processed
+                if channel.subscribed:
+                    if channel_dict["state"] == CHANNEL_STATE.UPDATING.value:
+                        try:
+                            progress = self.session.mds.compute_channel_update_progress(channel)
+                            channel_dict["progress"] = progress
```
[the remark this answers is not in the document]
Comment (ichorid) [reply src/tribler-core/tribler_core/modules/metadata_store/restapi/channels_endpoint.py:94]: No, it should not. We only describe the higher-level JSON scheme and do not dive into the Channels-specific details. This `progress` field is assigned directly to channel entries.

File: src/tribler-core/tribler_core/tests/test_notifier.py (hunk 1 above)
Comment (ichorid) [reply src/tribler-core/tribler_core/tests/test_notifier.py:19]: @qstokkink , thanks, I'll try that!

File: src/tribler-gui/tribler_gui/widgets/tablecontentmodel.py [hunk 5]
```diff
@@ -428,12 +449,11 @@ class SearchResultsModel(ChannelContentModel):
 
 
 class DiscoveredChannelsModel(ChannelContentModel):
-    columns = [u'state', u'votes', u'name', u'torrents', u'updated']
-    column_headers = [u'', tr('Popularity'), tr('Name'), tr('Torrents'), tr('Updated')]
-
-    column_width = {u'state': lambda _: 20, u'name': lambda table_width: table_width - 350}
+    columns = [u'subscribed', u'state', u'name', u'torrents', u'votes', u'updated']
+    column_headers = [u'Subscribed', u'', tr('Name'), tr('Torrents'), tr('Popularity'), tr('Updated')]
```
Comment (qstokkink) [inline src/tribler-gui/tribler_gui/widgets/tablecontentmodel.py:453]:
```suggestion
    column_headers = [tr('Subscribed'), '', tr('Name'), tr('Torrents'), tr('Popularity'), tr('Updated')]
```

File: src/tribler-gui/tribler_gui/widgets/triblertablecontrollers.py [hunk 6]
```diff
@@ -2,15 +2,22 @@
...
 
 from tribler_core.modules.metadata_store.serialization import CHANNEL_TORRENT, COLLECTION_NODE, REGULAR_TORRENT
 from tribler_core.utilities.json_util import dumps
 
+from tribler_gui.defs import HEALTH_CHECKING, HEALTH_UNCHECKED
 from tribler_gui.tribler_action_menu import TriblerActionMenu
 from tribler_gui.tribler_request_manager import TriblerNetworkRequest
+from tribler_gui.utilities import get_health
+
+HEALTHCHECK_DELAY = 500
```
Comment (xoriole) [inline src/tribler-gui/tribler_gui/widgets/triblertablecontrollers.py:20]: would be nice to mention the unit to avoid confusion eg.`# ms`

**Comment by qstokkink** [comment]
I like the new format. I made some small tweaks in resizing and text alignment to make it even more to my liking. Feel free to adopt or ignore any of these changes:

<details>
  <summary>Patch (click me to expand)</summary>
  
  ```patch
  diff --git a/src/tribler-gui/tribler_gui/widgets/tablecontentdelegate.py b/src/tribler-gui/tribler_gui/widgets/tablecontentdelegate.py
index 68b0bcd96..e39501759 100644
--- a/src/tribler-gui/tribler_gui/widgets/tablecontentdelegate.py
+++ b/src/tribler-gui/tribler_gui/widgets/tablecontentdelegate.py
@@ -453,7 +453,7 @@ class SubscribeToggleControl(QObject, CheckClickedMixin):
         self._margin = max(0, self._thumb_radius - self._track_radius)
         self._base_offset = max(self._thumb_radius, self._track_radius)
 
-        self._width = 4 * self._track_radius + 2 * self._margin
+        self._width = 6 * self._track_radius + 2 * self._margin
         self._height = 2 * self._track_radius + 2 * self._margin
 
         self._end_offset = {True: lambda: self._width - self._base_offset, False: lambda: self._base_offset}
diff --git a/src/tribler-gui/tribler_gui/widgets/tablecontentmodel.py b/src/tribler-gui/tribler_gui/widgets/tablecontentmodel.py
index 6bf0ff1a1..0e6b0071a 100644
--- a/src/tribler-gui/tribler_gui/widgets/tablecontentmodel.py
+++ b/src/tribler-gui/tribler_gui/widgets/tablecontentmodel.py
@@ -228,8 +228,8 @@ class ChannelContentModel(RemoteTableModel):
     }
 
     column_width = {
-        u'state': lambda _: 20,
-        u'subscribed': lambda _: 20,
+        u'state': lambda _: 200,
+        u'subscribed': lambda _: 100,
         u'name': lambda table_width: table_width - 510,
         u'action_buttons': lambda _: 20,
         u'category': lambda _: 20,
@@ -297,6 +297,8 @@ class ChannelContentModel(RemoteTableModel):
             return self.column_headers[num]
         if role == Qt.InitialSortOrderRole and num != self.column_position.get('name'):
             return Qt.DescendingOrder
+        elif role == Qt.TextAlignmentRole:
+            return Qt.AlignHCenter if num in [self.column_position.get('subscribed'), self.column_position.get('torrents')] else Qt.AlignLeft
 
     def rowCount(self, parent=QModelIndex()):
         return len(self.data_items)
@@ -359,6 +361,8 @@ class ChannelContentModel(RemoteTableModel):
         elif role == Qt.TextAlignmentRole:
             if index.column() == self.column_position.get(u'votes', -1):
                 return Qt.AlignLeft | Qt.AlignVCenter
+            if index.column() == self.column_position.get(u'torrents', -1):
+                return Qt.AlignHCenter | Qt.AlignVCenter
         return None
 
     def reset(self):
@@ -449,10 +453,10 @@ class SearchResultsModel(ChannelContentModel):
 
 
 class DiscoveredChannelsModel(ChannelContentModel):
-    columns = [u'subscribed', u'state', u'name', u'torrents', u'votes', u'updated']
-    column_headers = [u'Subscribed', u'', tr('Name'), tr('Torrents'), tr('Popularity'), tr('Updated')]
+    columns = [u'subscribed', u'name', u'state', u'torrents', u'votes', u'updated']
+    column_headers = [tr('Subscribed'), tr('Name'), '', tr('Torrents'), tr('Popularity'), tr('Updated')]
 
-    column_width = dict(ChannelContentModel.column_width, **{u'name': lambda table_width: table_width - 450})
+    column_width = dict(ChannelContentModel.column_width, **{u'name': lambda table_width: table_width - 630})
     default_sort_column = columns.index(u'votes')
 
     def __init__(self, *args, **kwargs):
  ```
  
</details>

**Comment by ichorid** [comment]
@qstokkink , I've checked your patch, it really makes the UI less cluttered. Though, I don't like the longer version of the subscribe toggle switch. These look as though they should be progress bars... Maybe I'll figure somethin out later.

File: src/tribler-gui/tribler_gui/widgets/channelcontentswidget.py (hunk 2 above)
Comment (ichorid) [reply src/tribler-gui/tribler_gui/widgets/channelcontentswidget.py:86]: Unfortunately, this does not work :disappointed:

File: src/tribler-core/tribler_core/modules/metadata_store/restapi/channels_endpoint.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ichorid) [reply src/tribler-core/tribler_core/modules/metadata_store/restapi/channels_endpoint.py:94]: Oops, my bad, you were talking about `ChannelSchema`, not the higher-level stuff... Will update that :+1:

File: src/tribler-core/tribler_core/modules/metadata_store/restapi/channels_endpoint.py [hunk 7]
```diff
@@ -79,7 +83,31 @@ async def get_channels(self, request):
...
+                        except (ZeroDivisionError, FileNotFoundError) as e:
```
[the remark this answers is not in the document]
Comment (ichorid) [reply src/tribler-core/tribler_core/modules/metadata_store/restapi/channels_endpoint.py:95]: Well, the situation where there are zero blobs should not ever happen... That's a good question, how to indicate it to the user?

**Comment by hbiyik** [comment]
i love this new PR, one suggestion dont you think hat it would be cool if there were a confirmation pop up in case unsubscribe is select off. it is somehow easy to click on those sliders while browsing.

**Comment by ichorid** [comment]
> i love this new PR, one suggestion dont you think hat it would be cool if there were a confirmation pop up in case unsubscribe is select off. it is somehow easy to click on those sliders while browsing.

Thanks for the suggestion! I was thinking about it. I'll probably file it later as a separate PR.

**Comment by qstokkink** [comment]
Very nice 👌

**Review by devos50 — COMMENTED** [review]
I checked out your changes and it already looks much better on Mac! 👍 

Still, I found a few pointers for improvement:

Schermafbeelding 2020-10-24 om 11 41 24

- There is still a 30px margin to the right of the table.

Schermafbeelding 2020-10-24 om 11 44 32

- I noticed three different visual styles in the toggle. Is this intended?
- Can you add a tooltip to the refresh button when browsing a specific channel?
- (Probably for another PR) I would find it more clear when the divider in the breadcrumb would have a different colour (e.g., white). Very minor suggestion but it might be easy to do.

Schermafbeelding 2020-10-24 om 11 48 48

- There is quite some margin to the left. I understand that the download button is there when you hover over the individual entries, but it looks a bit odd and inconsistent with the channels overview.

Schermafbeelding 2020-10-24 om 11 49 42

- There is also some padding to the right of the table when browsing the torrents in an individual channel.

I also asked a typical Tribler user that wishes to remain anonymous about her opinion. She found the user interface intuitive and clear. 👍

**Comment by ichorid** [comment]
@devos50 

>  I noticed three different visual styles in the toggle. Is this intended?

Yes, that's intended. The hollow version indicates that the channel is not downloaded/completed yet.

> Can you add a tooltip to the refresh button when browsing a specific channel?

:heavy_check_mark: Done.


>I would find it more clear when the divider in the breadcrumb would have a different colour (e.g., white).

:heavy_check_mark: Done

>There is quite some margin to the left. I understand that the download button is there ...

Adjusted the margin as best as I could. It will require another (small) refactoring of channels tables to make the columns sizes auto-adjusting. Also, if you have better ideas where to put the :arrow_down: download button - welcome! I'm out of ideas for it, at least for now.

**Review by devos50 — COMMENTED** [review]
At this point, I'm probably being too perfectionist but I still found a few things:

Schermafbeelding 2020-10-24 om 14 34 54
Schermafbeelding 2020-10-24 om 14 34 23

- There is still a bit of tiny margin to the right :(

Schermafbeelding 2020-10-24 om 14 34 43

- When opening a channel with no torrents, there is a random 'C' in the table header. When extending this column, it shows 'Category'.

**Comment by ichorid** [comment]
@devos50 , I deliberately left that "S" in place, so I could ask you about what to do with it. Really, should we remove this header?

Regarding the margin, I'm not going to fix it in this PR. The thing looks differently on different OSes. It will require refactoring the table model to fix it properly.

**Comment by devos50** [comment]
I think we can safely remove the 'Category' table header entry 👍

**Comment by devos50** [comment]
Also, don't forget to rebase your PR

**Review by devos50 — APPROVED** [review]

**Pull request merged**